rust/worker/src/execution/operators/filter.rs

```diff
@@ -0,0 +1,519 @@
+    fn code(&self) -> ErrorCodes {
+        use FilterError::*;
+        match self {
+            RecordSegmentReaderCreationError(e) => e.code(),
+            LogMaterializationError(e) => e.code(),
+            IndexError(e) => e.code(),
+            MetadataSegmentReaderError(e) => e.code(),
+            RecordSegmentReaderError => ErrorCodes::Internal,
+            InvalidInput => ErrorCodes::InvalidArgument,
+        }
+    }
+}
+
+/// This sturct provides an abstraction over the materialized logs that is similar to the metadata segment
+pub(crate) struct MetadataLogReader<'me> {
+    // This maps metadata keys to `BTreeMap`s, which further map values to offset ids
+    // This mimics the layout in the metadata segment
+    // //TODO: Maybe a sorted vector with binary search is more lightweight and performant?
+    compact_metadata: HashMap<&'me str, BTreeMap<&'me MetadataValue, RoaringBitmap>>,
+    // This maps offset ids to documents, excluding deleted ones
+    document: HashMap<u32, &'me str>,
+    // This contains all existing offset ids that are touched by the logs
+    touched_oids: RoaringBitmap,
+    // This maps user ids to offset ids, excluding deleted ones
+    uid_to_oid: HashMap<&'me str, u32>,
+}
+
+impl<'me> MetadataLogReader<'me> {
+    pub(crate) fn new(logs: &'me Chunk<MaterializedLogRecord<'me>>) -> Self {
+        let mut compact_metadata: HashMap<_, BTreeMap<&MetadataValue, RoaringBitmap>> =
+            HashMap::new();
+        let mut document = HashMap::new();
+        let mut touched_oids = RoaringBitmap::new();
+        let mut uid_to_oid = HashMap::new();
+        for (log, _) in logs.iter() {
+            if !matches!(
+                log.final_operation,
+                MaterializedLogOperation::Initial | MaterializedLogOperation::AddNew
+            ) {
+                touched_oids.insert(log.offset_id);
+            }
+            if !matches!(
+                log.final_operation,
+                MaterializedLogOperation::DeleteExisting
+            ) {
+                uid_to_oid.insert(log.merged_user_id_ref(), log.offset_id);
+                let log_meta = log.merged_metadata_ref();
+                for (key, val) in log_meta.into_iter() {
+                    compact_metadata
+                        .entry(key)
+                        .or_default()
+                        .entry(val)
+                        .or_default()
+                        .insert(log.offset_id);
+                }
+                if let Some(doc) = log.merged_document_ref() {
+                    document.insert(log.offset_id, doc);
+                }
+            }
+        }
+        Self {
+            compact_metadata,
+            document,
+            touched_oids,
+            uid_to_oid,
+        }
+    }
+    pub(crate) fn get(
+        &self,
+        key: &str,
+        val: &MetadataValue,
+        op: &PrimitiveOperator,
+    ) -> Result<RoaringBitmap, FilterError> {
+        use Bound::*;
```

Comment:
style question - why are we doing a import here? Can we just hoist this up? I generally find scoped imports like this a bit odd. Are we avoiding some namespace pollution?

Reply:
Sometimes the enum variants contain common keywords that may pollute the namespace, and I use the scoped imports as a way to limit the effects of these `use <enum>::*;` imports. For most of the time there is no collision though.

Reply:
I removed all scoped imports for clarity.